tests/integration/sharding_tests/application/lib/charms/mongos/v0/mongos_client_interface.py

```diff
@@ -0,0 +1,133 @@
+2. the host application receives the generated username, password and uri for connecting to the 
+sharded cluster.
+
+This library contains the Requires and Provides classes for handling the relation between an
+application and mongos. The mongos application relies on the MongosProvider class and the deployed
+application uses the MongoDBRequires class.
+
+The following is an example of how to use the MongoDBRequires class to specify the roles and 
+database name:
+
+```python
+from charms.mongos.v0.mongos_client_interface import MongosRequirer
+
+
+class ApplicationCharm(CharmBase):
+
+    def __init__(self, *args):
+        super().__init__(*args)
+
+        # relation events for mongos client
+        self._mongos_client = MongosRequirer(
+            self,
+            database_name="my-test-db",
+            extra_user_roles="admin",
+        )
+```
+
+To receive the username, password, and uri:
+# TODO this is to be implemented in a future PR
+"""
+
+
+class MongosProvider(Object):
+    """Manage relations between the mongos router and the application on the mongos side."""
+
+    def __init__(self, charm: CharmBase, relation_name: str = MONGOS_RELATION_NAME) -> None:
+        """Constructor for MongosProvider object."""
+        self.relation_name = relation_name
+        self.charm = charm
+        self.database_provides = DatabaseProvides(self.charm, relation_name=self.relation_name)
+
+        super().__init__(charm, self.relation_name)
+        self.framework.observe(
+            charm.on[self.relation_name].relation_changed, self._on_relation_changed
+        )
+
+        # TODO Future PRs handle relation broken
+
+    def _on_relation_changed(self, event) -> None:
+        """Handles updating the database and extra user roles."""
+        if not self.charm.unit.is_leader():
+            return
+
+        relation_data = event.relation.data[event.app]
+        new_database_name = relation_data.get(DATABASE_KEY, self.charm.database)
+        new_extra_user_roles = relation_data.get(USER_ROLES_KEY, self.charm.extra_user_roles)
+
+        if new_database_name != self.charm.database:
+            self.charm.set_database(new_database_name)
+
+        if new_extra_user_roles != self.charm.extra_user_roles:
+            if isinstance(new_extra_user_roles, str):
+                new_extra_user_roles = [new_extra_user_roles]
+
+            self.charm.set_user_roles(new_extra_user_roles)
+
+
+class MongosRequirer(Object):
+    """Manage relations between the mongos router and the application on the application side."""
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        database_name: str,
+        extra_user_roles: str,
+        relation_name: str = MONGOS_RELATION_NAME,
+    ) -> None:
+        """Constructor for MongosRequirer object."""
+        self.relation_name = relation_name
+        self.charm = charm
+
+        if not database_name:
+            database_name = f"{self.charm.app}"
```

Comment:
nits: 
maybe we can log here the fact that we are falling back to default db name?

Reply:
Hi not sure if you saw this in the description
> Feel free to ignore the test charm we copied over to tests/integration/sharding_tests/application/, this should be removed in the future

It was a test charm copied over from `mongos`